src/nifs.sol

```diff
@@ -0,0 +1,335 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity ^0.8.0;
+
+import "src/pasta/Pallas.sol";
+import "src/pasta/Vesta.sol";
+import "src/verifier/step2/Step2Data.sol";
+import "src/poseidon/Sponge.sol";
+
+library NIFSPallas {
+    uint256 constant private MOD = Pallas.P_MOD;
+    uint256 constant private NUM_FE_FOR_RO = 24; // <-- This seems strange, I'm only counting 15...
+
+    struct NIFS {
+        uint8[32] compressed_comm_T;
+    }
+
+    struct R1CSInstance {
+        Pallas.PallasAffinePoint comm_W;
+        uint256[] X;
+    }
+
+    struct RelaxedR1CSInstance {
+        Pallas.PallasAffinePoint comm_W;
+        Pallas.PallasAffinePoint comm_E;
+        uint256[] X;
+        uint256 u;
+    }
+
+    function verify(
```

Comment:
I would love a bit more detailed TODOs here for the steps that are missing (and that I agree we can add later).

Reply:
At this point we are simply failing on the test of the `verify` function. The verbose output from `forge test` is only moderately helpful in tracking down what assertion is failing.